drivers: video: ov2640 driver: Enhancing ov2640 driver

cornway opened the pull request:
Enabling proper zoom/scaling for smaller resolutions (e.g. QQVGA)
This is basically a copy paste from linux driver with some tweaks

This patch may need some refactoring/cleaning up, but generally I have verified it function properly
I found that without this patch, if i  enable low resolution format (e.g. QVGA/QQVGA) I basically receive cropped image (top left corner) not scaled, so it makes it difficult to receive meaningfull image from camera for further processing

josuah commented:
Thanks for the contribution!
However, Linux license is GPLv2 so it's not really possible to copy-paste from it.
Also, some possibly better way to handle it would be to allow the user to pass exactly any size they want.
Since it's computed from a macro, that might allow some runtime customization?

I think it supports arbitrary scaling, so this could be a good fit for the [`video_set_selection()`](https://docs.zephyrproject.org/latest/doxygen/html/group__video__interface.html#ga21f2e7d6b5ec0c50ceeee580c6272613) API to allow the user to crop/scale at any supported size.

dsseng commented on drivers/video/ov2640.c:
```diff
@@ -472,29 +568,23 @@ struct ov2640_data {
-	OV2640_VIDEO_FORMAT_CAP(1600, 1200, VIDEO_PIX_FMT_JPEG),   /* UXGA  */
-	{ 0 }
-};
+	OV2640_VIDEO_FORMAT_CAP(QQVGA_WIDTH, QQVGA_HEIGHT, VIDEO_PIX_FMT_RGB565), /* QQVGA */
+	OV2640_VIDEO_FORMAT_CAP(QCIF_WIDTH, QCIF_HEIGHT, VIDEO_PIX_FMT_RGB565),   /* QCIF  */
+	OV2640_VIDEO_FORMAT_CAP(CIF_WIDTH, CIF_HEIGHT, VIDEO_PIX_FMT_RGB565),     /* CIF   */
+	OV2640_VIDEO_FORMAT_CAP(VGA_WIDTH, VGA_HEIGHT, VIDEO_PIX_FMT_RGB565),     /* VGA   */
+	OV2640_VIDEO_FORMAT_CAP(SVGA_WIDTH, SVGA_HEIGHT, VIDEO_PIX_FMT_RGB565),   /* SVGA  */
+	OV2640_VIDEO_FORMAT_CAP(XGA_WIDTH, XGA_HEIGHT, VIDEO_PIX_FMT_RGB565),     /* XVGA  */
+	OV2640_VIDEO_FORMAT_CAP(SXGA_WIDTH, SXGA_HEIGHT, VIDEO_PIX_FMT_RGB565),   /* SXGA  */
+	OV2640_VIDEO_FORMAT_CAP(UXGA_WIDTH, UXGA_HEIGHT, VIDEO_PIX_FMT_RGB565),   /* UXGA  */
+	OV2640_VIDEO_FORMAT_CAP(QQVGA_WIDTH, QQVGA_HEIGHT, VIDEO_PIX_FMT_JPEG),   /* QQVGA */
+	OV2640_VIDEO_FORMAT_CAP(QCIF_WIDTH, QCIF_HEIGHT, VIDEO_PIX_FMT_JPEG),     /* QCIF  */
+	OV2640_VIDEO_FORMAT_CAP(CIF_WIDTH, CIF_HEIGHT, VIDEO_PIX_FMT_JPEG),       /* CIF   */
+	OV2640_VIDEO_FORMAT_CAP(VGA_WIDTH, VGA_HEIGHT, VIDEO_PIX_FMT_JPEG),       /* VGA   */
+	OV2640_VIDEO_FORMAT_CAP(SVGA_WIDTH, SVGA_HEIGHT, VIDEO_PIX_FMT_JPEG),     /* SVGA  */
+	OV2640_VIDEO_FORMAT_CAP(XGA_WIDTH, XGA_HEIGHT, VIDEO_PIX_FMT_JPEG),       /* XVGA  */
+	OV2640_VIDEO_FORMAT_CAP(SXGA_WIDTH, SXGA_HEIGHT, VIDEO_PIX_FMT_JPEG),     /* SXGA  */
+	OV2640_VIDEO_FORMAT_CAP(UXGA_WIDTH, UXGA_HEIGHT, VIDEO_PIX_FMT_JPEG),     /* UXGA  */
+	{0}};
```
Not sure if breaking these out to constants is wanted

josuah replied:
It can be hard to remember what size is UXGA or QCIF, while 1600x1200 or 176x144 is obvious for the reader.
Maybe the goal was avoiding accidental typos?

dsseng replied:
Then it would make sense to add numeric resolutions to the comments here, so they do not duplicate what's in the code

cornway replied:
That was done to match window setting, so if format is meant to be supported - it is known that there is a scaling/windowing setting for it
Thank you

dsseng reviewed (commented):
> This is basically a copy paste from linux driver with some tweaks

Please notice the ov2640 driver in Linux has no dual-licensing, and GPLv2 must not be copied into Zephyr. There seems to not be a significant code added, so let's wait for another review

P.S. Also please consider the SonarQube warnings

cornway commented:
Honestly, I just wanted to ease the pain for those still working with that type of sensors, it took me some time to figure this out

> Thanks for the contribution! However, Linux license is GPLv2 so it's not really possible to copy-paste from it. Also, some possibly better way to handle it would be to allow the user to pass exactly any size they want. Since it's computed from a macro, that might allow some runtime customization?
> 
> I think it supports arbitrary scaling, so this could be a good fit for the [`video_set_selection()`](https://docs.zephyrproject.org/latest/doxygen/html/group__video__interface.html#ga21f2e7d6b5ec0c50ceeee580c6272613) API to allow the user to crop/scale at any supported size.

One more note: there is no possibility (seems to me) of having "any size" user wants, as there are a bunch of prescalers and dividers to be set to downsize properly, not every size fits into that combination
Thank you for your patience and time

josuah commented on drivers/video/ov2640.c:
```diff
@@ -152,17 +152,114 @@ LOG_MODULE_REGISTER(video_ov2640, CONFIG_VIDEO_LOG_LEVEL);
+#define UXGA_HSIZE (1600UL)
+#define UXGA_VSIZE (1200UL)
+
+#define VAL_SET(x, mask, rshift, lshift) ((((x) >> rshift) & mask) << lshift)
```
Is this different from [`FIELD_PREP()`](https://docs.zephyrproject.org/latest/doxygen/html/group__sys-util.html#gaa03c8b31bf67a097dd9f8153a04ef86b)?

josuah replied:
There is no shift argument, but the macros can understand where the mask starts and ends, and shifts the value accordingly.

> representing its field **position** and width [[1]](https://docs.zephyrproject.org/latest/doxygen/html/group__sys-util.html#gaa03c8b31bf67a097dd9f8153a04ef86b)

To pre-compute the mask, you can use `GENMASK(lsb, msb)`, these two are typically used together and allow to use the value from the datasheet without risk of typo by manual tweaking the masks:

```c
#define THIS_FIELD_SET(x) FIELD_PREP(GENMASK(7, 3), x)
```

Or even better IMHO, just a plain transcription of the datasheet data without any smart:

```c
#define THIS_FIELD_MASK GENMASK(7, 3)
```

Then in the code:

```c
    {THIS_ADDR, FIELD_PREP(THIS_FIELD_MASK, 323)},
```

josuah commented on drivers/video/ov2640.c:
```diff
@@ -152,17 +152,114 @@ LOG_MODULE_REGISTER(video_ov2640, CONFIG_VIDEO_LOG_LEVEL);
+#define UXGA_HSIZE (1600UL)
+#define UXGA_VSIZE (1200UL)
+
+#define VAL_SET(x, mask, rshift, lshift) ((((x) >> rshift) & mask) << lshift)
```
Is that possible to use `FIELD_PREP()`? This would also break the chain from GPL code to Zephyr doing so.

josuah commented:
Ah, there we go, a WIP pull request showing some way to integrate cropping into an image sensor driver:
- https://github.com/zephyrproject-rtos/zephyr/pull/93797

josuah commented:
> -    I address comments regarding resolution constants
> -    Remove unnecessary changes (like code formatting), or make separate commit out of it
> -    Add this `if (ret < 0) { return ret; }`
> -    And most important - remove GPL'ed parts (e.g. reuse FIELD_PREP() macro)

I confirm that's all.
The improvements mentioned are long-term ideals.
I will propose some examples below.

Thank you again!

dsseng commented on drivers/video/ov2640.c:
```diff
@@ -346,61 +443,60 @@ static const struct ov2640_reg default_regs[] = {
+	{VSTOP, 0x97},  /* UXGA=0x97, SVGA/CIF=0x4b */
+	{0x3d, 0x34},   /* UXGA=0x34, SVGA/CIF=0x38 */
+
+	{0x35, 0x88},
+	{0x22, 0x0a},
+	{0x37, 0x40},
+	{0x34, 0xa0},
+	{0x06, 0x02},
+	{0x0d, 0xb7},
+	{0x0e, 0x01},
+	{0x42, 0x83},
```
Could you please name commits differently, to make sure the commit title is informative?

For example:
- 3355ed08f5c5c4a53152acb389c97d13a2e5cfb3 -> `drivers: video: ov2640: format code`
- 53964c7a13db25f78306cb75389901437bc7e433 -> `drivers: video: ov2640: add support for more resolutions`
- Please do not hesitate to change what I suggested, you know better what changes each commit contains and what might be of interest in it

Thanks!

Best,
Dmitrii

cornway replied:
Done

josuah reviewed (commented):
Would it be possible to do these two changes? This might be an occasion to get familiar with the Zephyr [doc on contributing](https://docs.zephyrproject.org/latest/contribute/index.html) PRs to be all set for subsequent PRs:

- In a PR, introducing something is expected to not be fixed by the next commits, so while it's appreciated that you kept the formatting changes separate, it will need to have the right formatting directly for newly contributed code.
- You might need to put different 1st line for the commits, to avoid confusion of which commit does what. Something very short is ok.
- We might need to keep 240x240 for now, or update the various samples using ESP32, as it is the default

Sorry for the late request for change, this PR did stay quite some time without review.

Thank you for your patience and for helping with image sensor driver maintenance!

cornway committed:
drivers: video: ov2640 driver: Enhancing ov2640 driver

Enabling proper zoom/scaling for windowed resolutions;
e.g. smaller than uxga
That makes camera's ISP to downscale frame to a desired resolution

Signed-off-by: Roman Pustobaiev <romanpustobayev@gmail.com>

cornway committed:
drivers: video: ov2640 driver: code formatting

Apply code formatting to some parts of the ov2640 driver

Signed-off-by: Roman Pustobaiev <romanpustobayev@gmail.com>

cornway commented:
I did my best, and hopefully I satisfied all the requested changes this time 😄 
Kindly ask to review once again

josuah commented on drivers/video/ov2640.c:
```diff
@@ -151,18 +151,100 @@ LOG_MODULE_REGISTER(video_ov2640, CONFIG_VIDEO_LOG_LEVEL);
+static const struct ov2640_win_size ov2640_supported_win_sizes[] = {
+	OV2640_SIZE(QQVGA_WIDTH, QQVGA_HEIGHT, ov2640_qqvga_regs),
+	OV2640_SIZE(QCIF_WIDTH, QCIF_HEIGHT, ov2640_qcif_regs),
+	OV2640_SIZE(240, 240, ov2640_qvga_regs),
```
This seems to still use the QVGA registers even though the resolution is announced as 240x240.

As a result, the capture sample works with the ESP32 S3 Eye, but instead of looking like this (before):

![PXL_20250917_093827260](https://github.com/user-attachments/assets/9b750a61-df45-474b-8f50-18e306c352f7)

It looks like this (after):

![PXL_20250917_093936553](https://github.com/user-attachments/assets/1e0376ca-54ef-400c-9e7d-53ff8389642b)

I first rebased your PR on top of https://github.com/zephyrproject-rtos/zephyr/pull/95958 then ran it with this:

```
west build -b esp32s3_eye/esp32s3/procpu samples/drivers/video/capture
```

cornway replied:
I see what the problem is
Let me fix it first and I let you know once this is done
Thank you for going through verification

cornway replied:
Pushed, please take a look
You may need to adjust these prescalers

```
static const struct ov2640_reg ov2640_240x240_regs[] = {
	OV2640_ZOOM_CONFIG(240, 240, 2, 2, 4),
};
```
`2, 2, 4` if that doesn't work, unfortunately I don't have that board so can't check

Thank you!

josuah reviewed (dismissed):
I rebased this on top of `main` and tested again:

Before: the test circle is not centered even though picture taken in front of the sensor (as you described)

![Image](https://github.com/user-attachments/assets/276d759d-7a7b-489f-9371-f34dfe07ce1e)

After: the test circle is now an ovale, wrong proportions (limitation where it's not possible to set the offset of the readout insofar

![Image](https://github.com/user-attachments/assets/b1985fe1-d179-4e4a-94fd-c8678e4e9747)

This is a driver/doc/hardware limitation for another.
Though, this is IMHO a big improvement over cropping, in particular for every other format.

More background in a follow-up discussion.

Thanks for the improvements (and for waiting), LGTM!

avolmat-st reviewed (changes requested):
Thanks for the update. My understanding is that this will pick a resolution in the list in order to apply it to the sensor and that resolution might not be the exact resolution asked by the application 

(checking for width >= and height >= instead of width == / height ==).
In such case, upon a set_fmt, the resolution actually applied might be different from the one requested.
This is ok, however the information should be given back to the application in such case by setting back the new resolution into the set_fmt fmt pointer.

josuah commented on drivers/video/ov2640.c:
```diff
@@ -151,18 +151,103 @@ LOG_MODULE_REGISTER(video_ov2640, CONFIG_VIDEO_LOG_LEVEL);
+#define OV2640_ZOOM_CONFIG(x, y, v_div, h_div, pclk_div)                                           \
+	{CTRLI,                                                                                    \
+	 CTRLI_LP_DP | FIELD_PREP(GENMASK(5, 3), v_div) | FIELD_PREP(GENMASK(2, 0), h_div)},       \
+		{ZMOW, FIELD_PREP(GENMASK(7, 0), (x) >> 2)},                                       \
```
Clang-format is very bad at indenting macros.

Maybe wrapping this in `/* clang-format off */` and `/* clang-format on */` to allow to keep it manually indented with the rest auto-indented?

Do you still wish to adjust this?

avolmat-st commented on drivers/video/ov2640.c:
```diff
@@ -775,6 +861,17 @@ static int ov2640_set_vertical_flip(const struct device *dev, int enable)
+			return &ov2640_supported_win_sizes[i];
+		}
+	}
+	return NULL;
```
Nit: can you ensure to keep an empty line before return statement at the end of a function ?

avolmat-st commented on drivers/video/ov2640.c:
```diff
@@ -784,23 +881,30 @@ static int ov2640_set_resolution(const struct device *dev,
+		LOG_ERR("Couldn't find window size for desired resolution setting");
+		return -EINVAL;
+	}
+	LOG_INF("Selected resolution %ux%u", win->width, win->height);
```
Maybe simply a LOG_DBG would be enough ?

avolmat-st reviewed (changes requested):
Thanks for the update. Only very very minor modifications if possible.

josuah reviewed (commented):
In complement to @avolmat-st feedback, some minor non-blocking points.

The driver is old, and this is incremental change that brings it closer to be a "modern Zephyr sensor", so not everything needs be fixed in this PR.
